### PR TITLE
Add endpoint for legacy systems to trigger beatmapset cover regeneration

### PR DESCRIPTION
**Pull request by nekodex**
'cuz oauth is hard and effort for systems that will be replaced or something

NOTE: update `.env` and change `SHARED_COOKIE_SECRET` => `SHARED_INTEROP_SECRET` before deploying

### CODE REVIEW

File: app/Http/Middleware/LegacyInterOpAuth.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    /**
+     * Handle an incoming request.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Closure  $next
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        if (!$request->hasHeader('X-LIO-Signature') || !presence($request->query('timestamp'))) {
```
Comment (nanaya) [inline app/Http/Middleware/LegacyInterOpAuth.php]: you mean `!present`

File: app/Http/Middleware/LegacyInterOpAuth.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            abort(403);
+        }
+
+        $timestamp = $request->query('timestamp');
```
Comment (nanaya) [inline app/Http/Middleware/LegacyInterOpAuth.php]: might as well assign this first before checking for presence.

File: app/Http/Middleware/LegacyInterOpAuth.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        $signature = $request->header('X-LIO-Signature');
+        $expected = hash_hmac('sha1', $request->fullUrl(), config('osu.legacy.shared_interop_secret'));
+
+        $diff = Carbon::createFromTimestamp($timestamp)->diffInSeconds();
```
Comment (nanaya) [inline app/Http/Middleware/LegacyInterOpAuth.php]: slightly simpler(?) with `->addMinutes(5)->isPast()`?

File: app/Http/Middleware/LegacyInterOpAuth.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+        if ($diff > 300 || $signature !== $expected) {
```
Comment (nanaya) [inline app/Http/Middleware/LegacyInterOpAuth.php]: something timing attack something `hash_equals`. (also in various other places)

File: app/Http/Controllers/LegacyInterOpController.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+namespace App\Http\Controllers;
+
+use App\Jobs\RegenerateBeatmapsetCover;
+use App\Models\Beatmapset;
+use Illuminate\Foundation\Bus\DispatchesJobs;
+use Illuminate\Routing\Controller as BaseController;
+
+class LegacyInterOpController extends BaseController
+{
+    use DispatchesJobs;
```
Comment (nanaya) [inline app/Http/Controllers/LegacyInterOpController.php]: update app base controller to not explode on missing `$session` and use that instead.

Comment (nanaya) [inline app/Http/Controllers/LegacyInterOpController.php]: I can't type.

Comment (nanaya) [inline app/Http/Controllers/LegacyInterOpController.php]: wait what why is it ended up as another review what why

Comment (nanaya) [inline app/Http/Controllers/LegacyInterOpController.php]: I can't type.

**Comment by nanaya** [comment]
also styleci exploded again?

File: app/Http/Middleware/LegacyInterOpAuth.php (hunk 3 above)
Comment (nekodex) [inline app/Http/Middleware/LegacyInterOpAuth.php]: This is deliberately allowing for 5mins on either side due to system clock drift still being a thing (apparently :v)

Comment (nanaya) [inline app/Http/Middleware/LegacyInterOpAuth.php]: who had the idea to default absolute value it (ノ･_･)ノ︵ ┻━┻

Comment (nekodex) [inline app/Http/Middleware/LegacyInterOpAuth.php]: Cuz difference is absolute? >.> (e.g a 5 degree difference can mean either +5 or -5)

File: app/Http/Controllers/LegacyInterOpController.php (hunk 5 above)
Comment (nekodex) [inline app/Http/Controllers/LegacyInterOpController.php]: Won't that add csrf checks and stuff?

File: app/Http/Middleware/LegacyInterOpAuth.php (hunk 3 above)
Comment (nanaya) [inline app/Http/Middleware/LegacyInterOpAuth.php]: that sounds more like tolerance or margin of error =|

File: app/Http/Controllers/LegacyInterOpController.php (hunk 5 above)
Comment (nanaya) [inline app/Http/Controllers/LegacyInterOpController.php]: but that's not where it's defined?

Comment (nanaya) [inline app/Http/Controllers/LegacyInterOpController.php]: I think it's always active unless you put it in the exception or update the middleware definitions.

Comment (nanaya) [inline app/Http/Controllers/LegacyInterOpController.php]: Also the impact of failing verification is just it resets the session.

File: app/Http/Controllers/LegacyInterOpController.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    public function regenerateBeatmapsetCovers($id)
+    {
+        $beatmapset = Beatmapset::findOrFail($id);
+
+        $job = (new RegenerateBeatmapsetCover($beatmapset))->onQueue('beatmap_processor');
+        $this->dispatch($job);
+
+        return response(['success' => true]);
```
Comment (nanaya) [inline app/Http/Controllers/LegacyInterOpController.php]: just return array

File: app/Http/Middleware/LegacyInterOpAuth.php [hunk 7]
```diff
@@ -0,0 +1,30 @@
+<?php
+
```
Comment (nanaya) [inline app/Http/Middleware/LegacyInterOpAuth.php]: missing license header.

File: app/Http/Middleware/LegacyInterOpAuth.php [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        $timestamp = $request->query('timestamp');
+        $diff = Carbon::createFromTimestamp($timestamp)->diffInSeconds();
+        $signature = $request->header('X-LIO-Signature');
+        $expected = hash_hmac('sha1', $request->fullUrl(), config('osu.legacy.shared_interop_secret'));
+
+        if (!present($signature) || !present($timestamp) || $diff > 300 || !hash_equals($expected, $signature)) {
```
Comment (nanaya) [inline app/Http/Middleware/LegacyInterOpAuth.php]: inverse the check to save some characters =D (`if(present($timestamp) && $diff < 300 && ...) { return $next...; } abort...;

**Comment by peppy** [comment]
💯

**Pull request merged**